Fixes Cluster worker shutdown/restart

MSP-Greg opened the pull request:
Hopefully fixes issues with correct handling of workers when:

1. Using phased-restart
2. externally SIGTERM'd workers

'Misbehaving' workers may not have been previously handled correctly.

Commit 'cluster.rb - fixup worker wait in Cluster, add Worker#term?' - relatively simple code that changes the handling of workers, making sure to trigger SIGKILL if `@options[:worker_shutdown_timeout]` has been exceeded, correct `wait` logic, and does not remove `dead?` workers from Cluster `@workers` array

Commit 'Adds two tests for worker SIGTERM/respawn and phased-restart' - adds two tests for the above.  They're somewhat brittle, and may fail, but not often.  They take about 30 seconds each.

I added the tests to current master and on top of PR #1892, and both of the new tests failed in every job.

See PR #1892 and Issue #1904 for additional discussion.

dentarg commented on test/test_integration.rb:
```diff
@@ -72,8 +72,12 @@ def start_forked_server(argv)
+    begin
+      Process.wait2(pid)
+    rescue Errno::ECHILD
+      15
```
`15`?

MSP-Greg replied:
@dentarg

Being a windows type, I can't test this locally.

The change above in `stop_forked_server` was done because I did see infrequent (and false) Travis failures from it.  But, it isn't really part of this PR, so I reverted it.

It is wrong (should be `[nil, 15]`), but given the 'green' builds, it shows how infrequent the failures were...  Thanks.

MSP-Greg commented:
Tests updated to make use of `after_worker_fork` hook to track worker creation.  Faster and more reliable...

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+      c.workers 2
+      c.worker_shutdown_timeout 2
+      c.app TestApps::SLEEP
+      c.after_worker_fork { |idx| workers_booted += 1 }
```
Why do we need a new mechanism for determining if the process has finished booting? We already have two methods.

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+    l = Puma::Launcher.new conf, :events => @events
+
+    t = Thread.new do
+      Thread.current.abort_on_exception = true
```
Don't need to set, this is already set on all thread in helper.rb

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+
+    Thread.kill worker0
+    Thread.kill worker1
+    http0   = nil
```
what's the point of setting all of these to nil?

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+    assert_empty old_waited, msg
+  end
+
+  def test_worker_phased_restart
```
These tests are exactly the same except for 1 line, could we clean that up a bit?

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+    l.stop
+    assert_kind_of Thread, t.join, "server didn't stop"
+
+    assert_operator (Time.now.to_f - start_time).round(2), :<, 32
```
maybe add together what you're looking for explicitly here: 

```ruby
assert_operator (Time.now.to_f - start_time).round(2), :<, (conf.worker_shutdown_timeout + 30)
```

I forgot, where does the 30 come from?

nateberkopec commented on test/test_integration.rb:
```diff
@@ -389,7 +389,207 @@ def test_no_zombie_children
+    msg = "old_pids #{old_pids.inspect}  new_pids #{new_pids.inspect}  old_waited #{old_waited.inspect}"
+    assert_equal 2, new_pids.length, msg
+    assert_equal 2, old_pids.length, msg
+    assert_equal new_pids, (new_pids - old_pids), "#{msg}\nBoth workers should be replaced"
```
imo this would be slightly clearer:

```ruby 
assert_empty new_pids & old_pids
```

nateberkopec commented on lib/puma/cluster.rb:
```diff
@@ -227,11 +220,18 @@ def check_workers(force=false)
-      pids = []
-      while pid = Process.waitpid(-1, Process::WNOHANG) do
-        pids << pid
+      @workers.reject! do |w|
```
This looks the same as the looped one, so we should extract a new method.

MSP-Greg commented:
@nateberkopec

I rebased and updated most of the items you mentioned.  Some of the work in the test file could be shared by other tests, saving that for another PR...
